src/Storage/Device/Local.php

```diff
@@ -119,26 +111,120 @@ public function upload(string $source, string $path, int $chunk = 1, int $chunks
         }
 
         if ($chunks === $chunksReceived) {
-            for ($i = 1; $i <= $chunks; $i++) {
-                $part = dirname($tmp).DIRECTORY_SEPARATOR.pathinfo($path, PATHINFO_FILENAME).'.part.'.$i;
-                $data = file_get_contents($part);
-                if (! $data) {
-                    throw new Exception('Failed to read chunk '.$part);
-                }
-
-                if (! file_put_contents($path, $data, FILE_APPEND)) {
-                    throw new Exception('Failed to append chunk '.$part);
-                }
-                \unlink($part);
+            $this->joinChunks($path, $chunks);
+
+            return $chunksReceived;
+        }
+
+        return $chunksReceived;
+    }
+
+    /**
+     * Upload Data.
+     *
+     * Upload file contents to desired destination in the selected disk.
+     * return number of chunks uploaded or 0 if it fails.
+     *
+     * @param  string  $source
+     * @param  string  $path
+     * @param  string  $contentType
+     * @param int chunk
+     * @param int chunks
+     * @param  array  $metadata
+     * @return int
+     *
+     * @throws \Exception
+     */
+    public function uploadData(string $data, string $path, string $contentType, int $chunk = 1, int $chunks = 1, array &$metadata = []): int
+    {
+        $this->createDirectory(\dirname($path));
+
+        if ($chunks === 1) {
+            if (! \file_put_contents($path, $data)) {
+                throw new Exception('Can\'t write file '.$path);
             }
-            \unlink($tmp);
+
+            return $chunks;
+        }
+        $tmp = \dirname($path).DIRECTORY_SEPARATOR.'tmp_'.\basename($path).DIRECTORY_SEPARATOR.\basename($path).'_chunks.log';
+
+        $this->createDirectory(\dirname($tmp));
```

Comment:
Can we clean the $tmp directory when complete?

Reply:
It's happening in Line number 191

Reply:
yep looks like it is not this directory, I am left after each transfer with an empty folder as 
tmp_{filenmae}

Reply:
Looking into it

Reply:
Added a change to remove the directory as well, previously it was only removing the log file. Test and let me know if it works @fogelito